horace_core/sqw/@sqw/func_eval.m

```diff
@@ -116,10 +129,175 @@
+    if ~isa(func_handle, 'function_handle')
+        error( ...
+            'SQW:func_eval:invalid_argument', ...
+            'Argument ''func_handle'' must be a function handle.\nFound ''%s''.', ...
+            class(func_handle) ...
+        );
+    end
+
+    keyval_def = struct('all', false, 'outfile', '', 'filebacked', false);
+    flag_names = {'all'};
+    parse_opts.flags_noneg = true;
+    parse_opts.flags_noval = true;
+    [~, keyval, ~] = parse_arguments( ...
+        varargin, keyval_def, flag_names, parse_opts ...
+    );
+
+    if ~iscell(pars)
+        pars = {pars};
+    end
+    opts.all_bins = keyval.all;
+
+    if ~islognumscalar(keyval.filebacked)
+        error( ...
+            'SQW:func_eval:invalid_argument', ...
+            ['Parameter ''filebacked'' must be a scalar logical value.\n', ...
+             'Found ''%s'' with size [%s].'], ...
+            class(keyval.filebacked), num2str(size(keyval.filebacked)) ...
+        );
+    end
+    opts.filebacked = keyval.filebacked;
+
+    if isempty(keyval.outfile)
+        opts.outfile = {};
+    elseif ~iscell(keyval.outfile)
+        opts.outfile = {keyval.outfile};
+    else
+        opts.outfile = keyval.outfile;
+    end
+
+    outfiles_empty = all(cellfun(@(x) isempty(x), opts.outfile));
+    if ~outfiles_empty && (numel(win) ~= numel(opts.outfile))
+        error( ...
+            'SQW:func_eval:invalid_arguments', ...
+            ['Number of outfiles specified must match number of input objects.\n' ...
+             'Found %i outfile(s), but %i sqw object(s).'], ...
+            numel(opts.outfile), numel(win) ...
+        );
+    end
+    if outfiles_empty && opts.filebacked
+        opts.outfile = gen_array_of_tmp_file_paths( ...
+            numel(win), 'horace_func_eval', tmp_dir() ...
+        );
+    end
+end
+
+
+function paths = gen_array_of_tmp_file_paths(nfiles, prefix, base_dir)
+    % Generate a cell array of paths for temporary files to be written to
+    % Format of the file names follows:
+    %   <prefix>_<UUID>_<counter_with_padded_zeros>.tmp
+    if nfiles < 1
+        error('CUT:cut_accumulate_data_', ...
+              ['Cannot create temporary file paths for less than 1 file.' ...
+               '\nFound %i.'], nfiles);
+    end
+    uuid = char(java.util.UUID.randomUUID());
+    counter_padding = floor(log10(nfiles)) + 1;
+    format_str = sprintf('%s_%s_%%0%ii.tmp', prefix, uuid, counter_padding);
+    paths = cell(1, nfiles);
+    for i = 1:nfiles
+        paths{i} = fullfile(base_dir, sprintf(format_str, i));
+    end
+end
+
+
+function write_sqw_with_out_of_mem_pix(sqw_obj, outfile)
+    % Write the given SQW object to the given file.
+    % The pixels of the SQW object will be derived from the image signal array
+    % and npix array, saving in chunks so they do not need to be held in memory.
+    %
+    pix = sqw_obj.data.pix;
+    npix = sqw_obj.data.npix;
+    img_signal = sqw_obj.data.s;
```

Comment:
does variance need to be set explicitly to zeros?

Reply:
I believe this is done in `write_out_of_mem_pix`